src/rust/engine/src/scheduler.rs

```diff
@@ -159,11 +190,34 @@ impl Session {
 pub struct ExecutionRequest {
   // Set of roots for an execution, in the order they were declared.
   pub roots: Vec<Root>,
+  // An ExecutionRequest with `poll` set will wait for _all_ of the given roots to have changed
+  // since their previous observed value in this Session before returning them.
+  //
+  // Example: if an ExecutionRequest is made twice in a row for roots within the same Session,
+  // and this value is set, the first run will request the roots and return immediately when they
+  // complete. The second request will check whether the roots have changed, and if they haven't
+  // changed, will wait until they have (or until the timeout elapses) before re-requesting them.
+  //
+  // TODO: The `poll`, `poll_delay`, and `timeout` parameters exist to support a coarse-grained API
```

Comment:
So, not going to lie: I'm not happy with how large this API ended up being. In a near-term followup, I'd like to move (the loop around) `run_goal_rules` into the rust `Scheduler`. But right now that would interfere with Greg's UI work.